### PR TITLE
Resolve #17

### LINKED ISSUE #17: Recursive continuation type crashes the interpreter
Using a recursive continuation type triggers an assertion failure in the interpreter. Minimal repo:
```wast
(module
  (type $ct (cont $ct))
  (func (param $k (ref $ct))
    (cont.bind $ct $ct (local.get $k))
    (drop)
  )
)
```

Output:
```
./wasm: uncaught exception File "valid/valid.ml", line 107, characters 16-22: Assertion failed
Raised at Wasm__Valid.func_type_of_heap_type in file "valid/valid.ml", line 107, characters 16-28
Called from Wasm__Valid.func_type_of_cont_type in file "valid/valid.ml" (inlined), line 112, characters 2-32
Called from Wasm__Valid.check_instr in file "valid/valid.ml", line 630, characters 27-59
Called from Wasm__Valid.check_seq in file "valid/valid.ml", line 1025, characters 26-43
Called from Wasm__Valid.check_block in file "valid/valid.ml", line 1030, characters 15-41
Called from Stdlib__List.iter in file "list.ml", line 112, characters 12-15
Called from Wasm__Valid.check_module in file "valid/valid.ml", line 1214, characters 2-42
Called from Wasm__Run.run_command in file "script/run.ml", line 518, characters 6-26
Called from Stdlib__List.iter in file "list.ml", line 112, characters 12-15
Called from Wasm__Run.run_script in file "script/run.ml" (inlined), line 588, characters 2-30
Called from Wasm__Run.run_quote_script in file "script/run.ml", line 593, characters 7-24
Re-raised at Wasm__Run.run_quote_script in file "script/run.ml", line 593, characters 58-67
Called from Wasm__Run.input_from in file "script/run.ml", line 104, characters 4-14
Called from Wasm__Run.input_sexpr_file in file "script/run.ml", line 144, characters 18-39
Re-raised at Wasm__Run.input_sexpr_file in file "script/run.ml", line 147, characters 27-36
Called from Wasm__Run.run_meta in file "script/run.ml", line 565, characters 16-50
Called from Stdlib__List.iter in file "list.ml", line 112, characters 12-15
Called from Wasm__Run.input_from in file "script/run.ml", line 104, characters 4-14
Called from Dune__exe__Wasm.(fun) in file "wasm.ml", line 45, characters 33-53
Called from Stdlib__List.iter in file "list.ml", line 112, characters 12-15
Called from Dune__exe__Wasm in file "wasm.ml", line 45, characters 4-72
```

### PR DESCRIPTION
**Pull request by dhil**
This patch (re)installs the check for continuation type well-formedness.

Currently, it allows only continuation types to be indexed by plain function types.

### CODE REVIEW

**Comment by dhil** [comment]
Should we allow indices which are recursive function groups too?

**Comment by rossberg** [comment]
It should allow anything that expands to a function type.

**Comment by dhil** [comment]
I revisited the code. I realised it already handles recursive groups, as the `func_type` internally expands the definition. I have added two more tests.

File: interpreter/valid/valid.ml [hunk 1]
```diff
@@ -183,7 +183,9 @@ let check_func_type (c : context) (ft : func_type) at =
 
 let check_cont_type (c : context) (ct : cont_type) at =
   match ct with
-  | ContT ft -> check_heap_type c ft at
+  | ContT (VarHT (StatX x)) ->
+     let _dt = func_type c (x @@ at) in ()
```
Comment (rossberg) [inline interpreter/valid/valid.ml]: Nit:
```suggestion
    let _dt = func_type c (x @@ at) in ()
```

File: interpreter/valid/valid.ml [hunk 2]
```diff
@@ -183,7 +183,9 @@ let check_func_type (c : context) (ft : func_type) at =
...
+  | _ -> error at "ill-formed continuation type"
```
Comment (rossberg) [inline interpreter/valid/valid.ml]:
```suggestion
  | _ -> assert false
```
In source, this cannot be anything else by construction. Analogous to how `sub` handles it.

**Review by rossberg — APPROVED** [review]

**Pull request merged**